Add OptionalRef; `clamp`: port to structured kernel

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61361

This PR ports the `clamp` kernel to the structured format. In addition, it introduces `OptionalScalarRef` as a replacement for `c10::optional<Scalar>&`. The latter, although it is a reference type, can still involve copying the contained `Scalar` (e.g. if the actual parameter is a `Scalar` or if a `c10::optional<Scalar>` is constructed just to call a kernel). `OptionalScalarRef` contains only a `const Scalar&`, and stores flag about whether the instance contains something inside the `Scalar` itself using a new tag.

For more information, see #55070.

Differential Revision: [D29821533](https://our.internmc.facebook.com/intern/diff/D29821533)

Commit:
[WIP] `clamp`: port to structured kernel

For more information, see #55070.

[ghstack-poisoned]

Commit:
Update on "[WIP] `clamp`: port to structured kernel"


For more information, see #55070.

[ghstack-poisoned]

Commit:
Update on "[WIP] `clamp`: port to structured kernel"


For more information, see #55070.

[ghstack-poisoned]

makslevental commented:
@SplitInfinity sorry just wondering: is this PR stalled? I'm kind of blocked on it (need out variant of `relu`). I looked at the CI logs and it looks like the error is due to some syntax in a file not touched by this diff:

![image](https://user-images.githubusercontent.com/5657668/125652356-fbee9b51-b9bb-478c-91cf-ea839896779d.png)


https://github.com/pytorch/pytorch/pull/61361/checks?check_run_id=3022080451#step:5:15930

but I don't think you've changed any macros that are used in that file.

SplitInfinity commented:
It is not stalled, I'm going to iterate on it and hopefully get all tests to pass today after submitting PSC>

makslevental commented:
> It is not stalled, I'm going to iterate on it and hopefully get all tests to pass today after submitting PSC.

great. lmk if need help getting it across the finish line

Commit:
Update on "[WIP] `clamp`: port to structured kernel"


For more information, see #55070.

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/runtime/static/ops.cpp:
```diff
@@ -340,7 +340,7 @@ REGISTER_OPERATOR_FUNCTOR(aten::clamp, aten_clamp, [](Node* n) -> SROperator {
       auto in1_s = p_node->Input(1).toOptional<at::Scalar>();
       auto in2_s = p_node->Input(2).toOptional<at::Scalar>();
-      at::native::clamp_out(in0_t, in1_s, in2_s, out_t);
+      out_t = at::clamp(in0_t, in1_s, in2_s);
```
This cannot be landed as is. How do I call the native version of the kernel I ported?

makslevental replied:
If I understand your question correctly, I think just `out_t = at::clamp_out(in0_t, in1_s, in2_s, out_t)` will work.

Compare with [thnn_conv2d_out](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L9387) which can be called `at::thnn_conv2d_out(out, x, weight, {1,1}, nullopt)`.

SplitInfinity replied:
Won't that still involve the dispatcher? Static runtime is supposed to not do that, right?

makslevental replied:
Sorry you're right. Hmm I guess one hacky solution would be

```cpp
    c10::optional<at::Tensor> in1_t;
    c10::optional<at::Tensor> in2_t;
    if (p_node->Input(1).isTensor()) {
      in1_t = p_node->Input(1).toOptional<at::Tensor>();
      in2_t = p_node->Input(2).toOptional<at::Tensor>();
    } else {
      in1_t = c10::make_optional(p_node->Input(1).toTensor());
      in2_t = c10::make_optional(p_node->Input(2).toTensor());
    }
    at::native::clamp_out(in0_t, in1_t, in2_t, out_t);
```

ezyang replied:
Don't do that: `at::cpu::clamp_out` is the name of the game. There are other occurrences of this pattern in this file too.

SplitInfinity commented on c10/core/Scalar.cpp:
```diff
@@ -32,5 +32,4 @@ Scalar Scalar::log() const {
     return std::log(v.i);
   }
 }
-
```
Undo.

SplitInfinity commented on c10/core/Scalar.h:
```diff
@@ -186,6 +195,27 @@ class C10_API Scalar {
+  }
+
+ private:
+  const Scalar empty_;
```
The purpose of this member is so that `scalar_` has something to refer to when the `OptionalScalarRef` instance refers to nothing. I tried to make it `static` in order to ensure there is only one instance per class and not per object, but I ran into linker errors I couldn't fix. I put the definition of said `static` member into `Scalar.cpp` and I suspect it there are some objects that include `Scalar.h` that aren't linked against whatever library `Scalar.cpp` ends up in. I dug through CMakeLists/Bazel files for a while but couldn't figure out how to fix this.

A fix for the above or another way to create something for `scalar_` to refer to when the instance of `OptionalScalarRef` refers to nothing would be appreciated. 🙏

ezyang replied:
Since your fields here are private, it seems better to just store a `Scalar*`. Then we don't even need `HAS_n`.

SplitInfinity commented on c10/core/Scalar.h:
```diff
@@ -24,6 +24,11 @@ namespace c10 {
 class C10_API Scalar {
  public:
+  // constructor tag used by empty scalar constructor
+  struct MakeEmptyScalar {};
+
+  explicit Scalar(MakeEmptyScalar) : tag(Tag::HAS_n) {}
```
I borrowed the tagged constructor strategy from `intrusive_ptr` 😋 

On second thought, these really should be private, and `OptionalScalarRef` should be a `friend` of `Scalar`.

SplitInfinity commented on c10/core/Scalar.h:
```diff
@@ -147,6 +152,10 @@ class C10_API Scalar {
     }
   }
 
+  bool hasValue() const {
+    return Tag::HAS_n != tag;
+  }
```
This should be private too.

ezyang commented on c10/core/Scalar.h:
```diff
@@ -176,7 +185,7 @@ class C10_API Scalar {
   // syntax v{ .member = ... } because it doesn't work on MSVC
 
-  enum class Tag { HAS_d, HAS_i, HAS_z, HAS_b };
+  enum class Tag { HAS_d, HAS_i, HAS_z, HAS_b, HAS_n };
```
`HAS_n` is a weird name for saying it doesn't have anything :P Worth a comment, or Note reference, here.

ezyang commented:
This looks substantively fine but seeing how the OptionalScalarRef implementation turned out I'm retracting my suggestion to you that a `HAS_n` tag would be good; you should just represent the nullary state with nullptr. I'm going to go ahead and approve this since I'm PTO tomorrow.

ezyang reviewed (approved):
please see comments

SplitInfinity commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -539,18 +550,23 @@ std::tuple<Tensor&, Tensor&> min_out(
+    clamp_scalar_stub(device_type(), *this, min.toScalar(), max.toScalar());
   } else if (max) {
-    at::clamp_max_outf(self, *max, result);
+    Tensor temp(result);
+    at::clamp_max_outf(self, max.toScalar(), temp);
+    result.copy_(temp);
```
Oh, this is another snippet I wanted to talk about. I would like to pass in `result` directly, but can't do that since the last parameter of `at::clamp_max_outf` is not `const`. Can this be fixed once `clamp_max` is ported to structured? Should I stack that + the fix for this on top?

SplitInfinity replied:
This kind of looks like #50953.

ezyang replied:
Do a const cast

ezyang replied:
(Don't `Tensor temp(result)`, that's an unnecessary refcount bump). In the terminal state `clamp_min_outf` takes a const argument and the const cast is unnecessary)

Commit:
Update on "`clamp`: port to structured kernel"


This PR ports the `clamp` kernel to the structured format. In addition, it introduces `OptionalScalarRef` as a replacement for `c10::optional<Scalar>&`. The latter, although it is a reference type, can still involve copying the contained `Scalar` (e.g. if the actual parameter is a `Scalar` or if a `c10::optional<Scalar>` is constructed just to call a kernel). `OptionalScalarRef` contains only a `const Scalar&`, and stores flag about whether the instance contains something inside the `Scalar` itself using a new tag.

For more information, see #55070.

[ghstack-poisoned]

Commit:
Update on "`clamp`: port to structured kernel"


This PR ports the `clamp` kernel to the structured format. In addition, it introduces `OptionalScalarRef` as a replacement for `c10::optional<Scalar>&`. The latter, although it is a reference type, can still involve copying the contained `Scalar` (e.g. if the actual parameter is a `Scalar` or if a `c10::optional<Scalar>` is constructed just to call a kernel). `OptionalScalarRef` contains only a `const Scalar&`, and stores flag about whether the instance contains something inside the `Scalar` itself using a new tag.

For more information, see #55070.

[ghstack-poisoned]

SplitInfinity commented on c10/util/Optional.h:
```diff
@@ -587,13 +587,16 @@ class optional : private OptionalBase<T> {
     return OptionalBase<T>::initialized();
   }
+
+ public:
```
I had to make `dataptr()` public now that `OptionalScalarRef` stores `const Scalar*`. But it doesn't seem right to do this.

Commit:
Update on "`clamp`: port to structured kernel"


This PR ports the `clamp` kernel to the structured format. In addition, it introduces `OptionalScalarRef` as a replacement for `c10::optional<Scalar>&`. The latter, although it is a reference type, can still involve copying the contained `Scalar` (e.g. if the actual parameter is a `Scalar` or if a `c10::optional<Scalar>` is constructed just to call a kernel). `OptionalScalarRef` contains only a `const Scalar&`, and stores flag about whether the instance contains something inside the `Scalar` itself using a new tag.

For more information, see #55070.

[ghstack-poisoned]

SplitInfinity commented:
Rerequesting review because switching to `Scalar*` required making `optional::dataptr()` `public` and that seems imprudent.

Commit:
Update on "`clamp`: port to structured kernel"


This PR ports the `clamp` kernel to the structured format. In addition, it introduces `OptionalScalarRef` as a replacement for `c10::optional<Scalar>&`. The latter, although it is a reference type, can still involve copying the contained `Scalar` (e.g. if the actual parameter is a `Scalar` or if a `c10::optional<Scalar>` is constructed just to call a kernel). `OptionalScalarRef` contains only a `const Scalar&`, and stores flag about whether the instance contains something inside the `Scalar` itself using a new tag.

For more information, see #55070.

[ghstack-poisoned]

ezyang commented on c10/core/Scalar.h:
```diff
@@ -186,6 +187,29 @@ class C10_API Scalar {
+
+  operator bool() const {
+    return has_value();
+  }
```
btw, if you hate optional's conversion to boolean, you could delete this implicit conversion and force people to write `has_value()` explicitly.

Commit:
Update on "`clamp`: port to structured kernel"


This PR ports the `clamp` kernel to the structured format. In addition, it introduces `OptionalScalarRef` as a replacement for `c10::optional<Scalar>&`. The latter, although it is a reference type, can still involve copying the contained `Scalar` (e.g. if the actual parameter is a `Scalar` or if a `c10::optional<Scalar>` is constructed just to call a kernel). `OptionalScalarRef` contains only a `const Scalar&`, and stores flag about whether the instance contains something inside the `Scalar` itself using a new tag.

For more information, see #55070.

[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/native/cuda/TensorCompare.cu:
```diff
@@ -60,7 +60,7 @@ void clamp_kernel_impl(TensorIterator& iter) {
 }
 
-void clamp_scalar_kernel_impl(TensorIterator& iter, Scalar min, Scalar max) {
+void clamp_scalar_kernel_impl(TensorIteratorBase& iter, Scalar min, Scalar max) {
```
Hold on, shouldn't these be `Scalar&` to avoid copying `Scalars` around?

Commit:
Update on "`clamp`: port to structured kernel"


This PR ports the `clamp` kernel to the structured format. In addition, it introduces `OptionalScalarRef` as a replacement for `c10::optional<Scalar>&`. The latter, although it is a reference type, can still involve copying the contained `Scalar` (e.g. if the actual parameter is a `Scalar` or if a `c10::optional<Scalar>` is constructed just to call a kernel). `OptionalScalarRef` contains only a `const Scalar&`, and stores flag about whether the instance contains something inside the `Scalar` itself using a new tag.

For more information, see #55070.

[ghstack-poisoned]

Commit:
Update on "`clamp`: port to structured kernel"


This PR ports the `clamp` kernel to the structured format. In addition, it introduces `OptionalScalarRef` as a replacement for `c10::optional<Scalar>&`. The latter, although it is a reference type, can still involve copying the contained `Scalar` (e.g. if the actual parameter is a `Scalar` or if a `c10::optional<Scalar>` is constructed just to call a kernel). `OptionalScalarRef` contains only a `const Scalar&`, and stores flag about whether the instance contains something inside the `Scalar` itself using a new tag.

For more information, see #55070.

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29821533).

ezyang commented on c10/core/OptionalRef.h:
```diff
@@ -0,0 +1,31 @@
+class OptionalRef {
+ public:
+  OptionalRef() : data_(nullptr) {}
+  OptionalRef(const T* data) : data_(data) {
```
I'd probably just get rid of this constructor at this point.

Commit:
Update on "Add OptionalRef; `clamp`: port to structured kernel"


This PR ports the `clamp` kernel to the structured format. In addition, it introduces `OptionalScalarRef` as a replacement for `c10::optional<Scalar>&`. The latter, although it is a reference type, can still involve copying the contained `Scalar` (e.g. if the actual parameter is a `Scalar` or if a `c10::optional<Scalar>` is constructed just to call a kernel). `OptionalScalarRef` contains only a `const Scalar&`, and stores flag about whether the instance contains something inside the `Scalar` itself using a new tag.

For more information, see #55070.

Differential Revision: [D29821533](https://our.internmc.facebook.com/intern/diff/D29821533)

[ghstack-poisoned]
